File: metal/app/Services/CheckoutService.php
<?php

namespace App\Services;

use App\Models\Order;
use App\Models\OrderItem;
use Illuminate\Support\Facades\DB;
use MercadoPago\Item;
use MercadoPago\Preference;

class CheckoutService
{
    private $productService;

    public function __construct(ProductService $productService)
    {
        $this->productService = $productService;
    }

    public function createOrder($order, $orderItem)
    {
        $product = $this->productService->get($orderItem['product_id']);
        $price = $this->productService->getPriceByMerchantTypeAttribute($orderItem['merchant_type_attribute_id']);

        $item = new Item();
        $item->title = $product->name;
        $item->quantity = $orderItem['quantity'] ?? 1;
        $item->picture_url = $product->thumb_url;
        $item->unit_price = $price;
        $item->currency_id = 'BRL';
        $preference = new Preference();
        $preference->items = [$item];
        $preference->auto_return = 'approved';
        $preference->back_urls = (object) [
            'success' => 'https://imprima.app/pagamento/sucesso',
            'pending' => 'https://imprima.app/pagamento/pendente',
            'failure' => 'https://imprima.app/pagamento/falha'
        ];
        $preference->save();

        $this->saveOrder($preference->id, $order, $orderItem, $price);

        return $preference;
    }

    private function saveOrder($preferenceId, $order, $orderItem, $price)
    {
        $sorder = new Order();
        $sorder->name = $order['name'];
        $sorder->cellphone = $order['cellphone'];
        $sorder->address = $order['address'];
        $sorder->note = $order['note'];
        $sorder->total = $orderItem['quantity'] * $price;
        $sorder->merchant_id = $order['merchant_id'];
        $sorder->origin = $order['origin'];
        $sorder->preference_id = $preferenceId;
        $sorder->save();

        $sorderItem = new OrderItem();
        $sorderItem->name = $orderItem['name'];
        $sorderItem->description = $orderItem['description'];
        $sorderItem->quantity = $orderItem['quantity'];
        $sorderItem->price = $price;
        $sorderItem->product_id = $orderItem['product_id'];
        $sorderItem->order_id = $sorder->id;
        $sorderItem->save();

        return $sorder;
    }

    public function updateOrder($preferenceId, $paymentType, $status)
    {
        $order = Order::where('preference_id', '=', $preferenceId)->first();
        $order->payment_type = $paymentType;
        $order->status = $status;
        $order->save();

        $merchant = DB::table('merchants')->find($order->merchant_id);
        return [$order, $merchant];
    }
}

File: plata/backend/app/app/service/checkout_service.py
from fastapi import APIRouter, Depends, Request
from sqlalchemy.orm import Session

from typing import Any, List

from app import crud, schemas
from app.api import deps

import base64
from uniqid import uniqid
from mercadopago.sdk import SDK


class CheckoutService:
    def save_image(self, base64_file):
        base_dir = "app/static/userspace/orders/"
        file = uniqid()
        file_out = base_dir+file+".png"
        with open(file_out, "wb+") as f:
            f.write(base64.b64decode(base64_file.replace("data:image/png;base64,", "")))

        return "userspace/orders/"+file+".png"

    def create_order(self, db: Session, order, order_items):

        sdk = SDK(
            "TEST-7463350488822882-101511-f6ea8328b793fa3a80b77e8ccaf1c1d9__LC_LA__-35226241")

        print(order, order_items)
        items = []
        total = 0
        for order_item in order_items:
            tmerchant_type_attribute = crud.merchant_type_attribute.get_by_id(
                db, order_item.id)
            item = {
                "title": order.product_name + ' ' + order_item.detail,
                "quantity": order_item.quantity,
                "unit_price": float(tmerchant_type_attribute.price),
                "currency_id": "BRL"
            }
            items.append(item)
            total = total + (order_item.quantity * tmerchant_type_attribute.price)

        preference_data = {
            "items": items,
            "auto_return": "approved",
            "back_urls": {
                "success": "https://imprima.app/checkout/payment/success",
                "pending": "https://imprima.app/checkout/payment/pending",
                "failure": "https://imprima.app/checkout/payment/fail"
            }
        }

        response = sdk.preference().create(preference_data)
        preference = response["response"]
        print(preference)

        porder = crud.order.create_order(db, preference["id"], order, total)

        for order_item in order_items:
            crud.order_item.create_order_item(
                db, porder.id, order_item)

        return preference

    def update_order(self, db, preference_id, payment_id, payment_type, status):
        db_order = crud.order.get_by_preference_id(db, preference_id)
        db_order.payment_id = payment_id
        db_order.payment_type = payment_type
        db_order.status = status
        db.commit()

        merchant = crud.merchant.get_by_id(db, db_order.merchant_id)
        return (db_order, merchant)
